partiql-spi/src/main/java/org/partiql/eval/value/DatumComparator.java

```diff
@@ -0,0 +1,720 @@
+package org.partiql.eval.value;
+
+import org.jetbrains.annotations.NotNull;
+import org.partiql.types.PType;
+
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+/**
+ * This class allows for the comparison between two {@link Datum}s. This is internally implemented by constructing
+ * a comparison table, where each cell contains a reference to a {@link DatumComparison} to compute the comparison.
+ * The table's rows and columns are indexed by the {@link PType.Kind#ordinal()}. The first dimension matches the
+ * left-hand-side's type of {@link #compare(Datum lhs, Datum rhs)}. The second dimension matches the right-hand-side's
+ * type of {@link #compare(Datum lhs, Datum rhs)}. As such, this implementation allows for O(1) comparison of scalars.
+ */
+abstract class DatumComparator implements Comparator<Datum> {
+
+    private static final int EQUAL = 0;
+
+    private static final int LESS = -1;
+
+    private static final int GREATER = 1;
+
+    @NotNull
+    private static final PType.Kind[] TYPE_KINDS = PType.Kind.values();
+
+    private static final int TYPE_KINDS_LENGTH = TYPE_KINDS.length;
+
+    /**
+     * This array defines the precedence of types when comparing values of different types.
+     * The lower the index, the higher the precedence. However, this is used for generating the BASELINE precedence.
+     * This may be overridden by families of types. For example, all number values are compared by their mathematical
```

Comment:
Ah, yeah. I was having trouble finding the right wording for this.

In this implementation, I initialize the two-dimensional array (see `initializeComparatorArray()`) by only performing comparisons between any two type-family's precedence.. So, for the cell representing the comparison between INT and STRING, I initialize the cell to return LESS (due to the numbers coming before text types). Similarly, for INT and BIGINT, I initialize the cell to return EQUALS (due to the shared type precedence). Note that this is only on the initialization of the array.

Now, in order to handle the comparisons *within* types of the same precedence, I need to manually rewrite some of those comparisons (the ones where there originally returned an EQUALS). So, for example, `fillBigIntComparator` is invoked before the static 2D table is returned, and it specifically handles cases such as the comparison of BIGINT and INT. It'll place within the table a function that compares the underlying JVM values of the BIGINT and INT.

So, with both of these passes, we can ensure that we have memoized the comparisons of values by handling both the overarching type family comparisons and the value comparisons (where necessary).

I'll update this specific Javadoc since it probably causes more confusion than necessary and will include comments in other private methods for better clarification.

Reply:
Thanks for the explanation. The updated javadoc is clearer for what's going on